typed/validate.go

```diff
@@ -152,11 +152,10 @@ func (v *validatingObjectWalker) visitListItems(t schema.List, list *value.List)
 			// this element.
 			continue
 		}
-		keyStr := pe.String()
-		if _, found := observedKeys[keyStr]; found {
-			errs = append(errs, v.errorf("duplicate entries for key %v", keyStr)...)
+		if observedKeys.Has(pe) {
```

Comment:
Is this really faster? i know the benchmarks are better but it makes the constant time operation with something that has to sort a slice and do a binary search

https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/fieldpath/element.go#L243-L255

Reply:
cc @lavalamp 

It is slightly faster but most importantly it's more correct. Using the string as the index is a little wrong since it's not 100% canonicalized and is not meant to be used for "production use".

Reply:
Allocating a buffer for the string, converting the pathelement to a string, hashing the string, comparing the strings, is quite expensive too

Reply:
String() was actually written to make this work, so this should be equally correct.

The set impl is O(log n) but with a low enough constant factor that it was faster than the O(1) thing it replaced. Now we don't have a O(1) thing to compare it with, I assume at some size it would stop being faster.